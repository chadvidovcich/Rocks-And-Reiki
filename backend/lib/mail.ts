/* eslint-disable @typescript-eslint/restrict-template-expressions */
import { createTransport, getTestMessageUrl } from 'nodemailer';
import Mail from 'nodemailer/lib/mailer';

const transport = createTransport({
  host: process.env.MAIL_HOST,
  port: Number(process.env.MAIL_PORT),
  auth: {
    user: process.env.MAIL_USER,
    pass: process.env.MAIL_PASS,
  },
});

function makeEmail(text: string): string {
  return `
  <div style="
  border: 1px solid black;
  padding: 20px;
  font-family: sans-serif
  line-height: 2
  font-size: 20px
  ">
  <h2>Hello There!</h2>
  <p>${text}</p>
  <p>üéâ, Chad</p>
  </div>
  `;
}

// eslint-disable-next-line import/prefer-default-export
export async function sendPasswordResetEmail(
  resetToken: string,
  to: string
): Promise<void> {
  // email the user a password reset token
  const info = await transport.sendMail({
    to,
    from: 'test@example.com',
    subject: 'Your password reset token!',
    html: makeEmail(`Your password reset token is here!
    <a href="${process.env.FRONTEND_URL}/reset?token=${resetToken}">Click Here to Reset</a>
    `),
  });

  if (process.env.MAIL_USER?.includes('ethereal.email')) {
    console.log(`‚úâÔ∏è Message Sent! Preview it at ${getTestMessageUrl(info)}`);
  }
}
