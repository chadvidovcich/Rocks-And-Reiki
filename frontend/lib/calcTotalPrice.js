export default function calcTotalPrice(cart) {
  return cart.reduce((sum, cartItem) => {
    if (!cartItem.product) return sum; //products could be deleted, but still in a user's cart
    return sum + cartItem.quantity * cartItem.product.price;
  }, 0);
}
