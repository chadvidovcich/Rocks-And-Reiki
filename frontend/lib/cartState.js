import { createContext, useContext, useState } from 'react';

const LocalStateContext = createContext();
const LocalStateProvider = LocalStateContext.Provider;

function CartStateProvider({ children }) {
  // this is our custom provider. we will store data (state) and functionality (updates) in here. Anyone can access it via the consumer

  const [cartOpen, setCartOpen] = useState(false);

  function toggleCart() {
    setCartOpen(!cartOpen);
  }

  function closeCart() {
    setCartOpen(false);
  }

  function openCart() {
    setCartOpen(true);
  }

  return (
    <LocalStateProvider
      value={{ cartOpen, setCartOpen, toggleCart, closeCart, openCart }}
    >
      {children}
    </LocalStateProvider>
  );
}

// make a custom hook for accessing the cart local state
function useCart() {
  const all = useContext(LocalStateContext);
  return all;
}

export { CartStateProvider, useCart };
