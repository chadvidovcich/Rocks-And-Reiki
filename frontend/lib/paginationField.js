import { PAGINATION_QUERY } from '../components/Pagination';

export default function paginationField() {
  return {
    keyArgs: false,
    read(existing = [], { args, cache }) {
      const { skip, first } = args;

      // read the number of items on the page
      const data = cache.readQuery({ query: PAGINATION_QUERY });
      const count = data?._allProductsMeta?.count;
      const page = skip / first + 1;
      const pages = Math.ceil(count / first);

      // check if we have existing items
      const items = existing.slice(skip, skip + first).filter((x) => x);
      // there are items on the last page, but not enough to fill a page fully
      if (items.length && items.length !== first && page === pages) {
        return items;
      }
      // we don't have any items. Need to fetch from network
      if (items.length !== first) {
        return false;
      }
      // we have items. return them from cache
      if (items.length) {
        console.log(
          `There are ${items.length} items in cache. Sending them to Apollo`
        );
        return items;
      }

      return false; // fallback to network
    },
    merge(existing = [], incoming, { args }) {
      const { skip } = args;
      console.log(`Merging ${incoming.length} items from the network`);
      const merged = existing ? existing.slice(0) : [];
      for (let i = skip; i < skip + incoming.length; ++i) {
        merged[i] = incoming[i - skip];
      }
      return merged;
    },
  };
}
