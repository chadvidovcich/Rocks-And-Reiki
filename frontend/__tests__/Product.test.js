// @ts-ignore
import { render, screen } from '@testing-library/react';
import Product from '../components/Product';
import { fakeItem } from '../lib/testUtils';
// @ts-ignore
import { MockedProvider } from '@apollo/react-testing';
const product = fakeItem();

describe('<Product/>', () => {
  it('renders out the price tag and title', () => {
    const { container, debug } = render(
      <MockedProvider>
        <Product product={product} />
      </MockedProvider>
    );

    const priceTag = screen.getByText('$50');
    expect(priceTag).toBeInTheDocument();
    const link = container.querySelector('a');
    expect(link).toHaveAttribute('href', `/product/${product.id}`);
    expect(link).toHaveTextContent(product.name);
  });

  it('Renders and matches the snapshot', () => {
    const { container, debug } = render(
      <MockedProvider>
        <Product product={product} />
      </MockedProvider>
    );
    expect(container).toMatchSnapshot();
  });

  it('Renders the image properly', () => {
    const { container, debug } = render(
      <MockedProvider>
        <Product product={product} />
      </MockedProvider>
    );
    const img = screen.getByAltText(product.name);
    expect(img).toBeInTheDocument();
  });
});
