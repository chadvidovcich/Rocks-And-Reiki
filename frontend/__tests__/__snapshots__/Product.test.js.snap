// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Product/> renders and matches the snapshot 1`] = `
<div>
  <div
    class="ItemStyles-sc-496aoe-0 cEPPbX"
  >
    <img
      alt="dogs are best"
      src="dog.jpg"
    />
    <h3
      class="Title-sc-1v5zspa-0 eMdfOY"
    >
      <a
        href="/product/abc123"
      >
        dogs are best
      </a>
    </h3>
    <span
      class="PriceTag-sc-1w27jzp-0 jxJElQ"
    >
      $50
    </span>
    <p>
      dogs
    </p>
    <div
      class="buttonList"
    >
      <a
        href="/update?id=abc123"
      >
        Edit 🤔🖊️
      </a>
      <button
        type="button"
      >
        Add
         To Cart🛒
      </button>
      <button
        type="button"
      >
        Delete
      </button>
    </div>
  </div>
</div>
`;
